examples/vision/adamatch.py

```diff
@@ -0,0 +1,453 @@
+            mask = tf.reduce_max(y_tilde_target_w, axis=-1) >= c_tau
+
+            ## Compute losses (pay attention to the indexing) ##
+            source_loss = compute_loss_source(
+                source_labels,
+                final_source_logits[: tf.shape(source_w)[0]],
+                final_source_logits[tf.shape(source_w)[0] :],
+            )
+            target_loss = compute_loss_target(
+                y_tilde_target_w, logits_target[tf.shape(target_w)[0] :], mask
+            )
+
+            t = mu_schedule()
+            total_loss = source_loss + (t * target_loss)
+            global T
+            T += 1
+
+        gradients = tape.gradient(total_loss, self.model.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.model.trainable_variables))
+
+        self.loss_tracker.update_state(total_loss)
+        return {"loss": self.loss_tracker.result()}
+
+
+"""
+The authors introduce three improvements in the paper:
+
+* In AdaMatch, we perform two forward passes and only one of them is respsonsible for
+updating the Batch Normalization statistics. This is done to account for the distribution
+shifts from the target dataset. In the other forward pass, we only use the source sample
+and the Batch Normalization layers are run in inference mode. Logits for the source
+samples (weakly and strongly augmented versions) from these two passes are slightly
+different from one another because of how Batch Normalization layers are run. Final
+logits for the source samples are computed by linearly interpolating between these two
+different pairs of logits. This induces a form of consistency regularization. This step
+is referred to as **random logit interpolation**.
+* **Distribution alignment** is used to align the source and target label distributions.
+This further helps the underlying model learn *domain-invariant representations*. In case
+of unsupervised domain adaptation, we don't have access to any labels of the target
+dataset. This is why pseudo labels are generated from the underlying model. 
+* The underlying model generates pseudo-labels for the target samples. It's likely that
+the model would make faulty predictions. Those can propagate back as we make progress in
+the training and hurt the overall performance. To compensate for that, we filter the
+high-confidence predictions based on a threshold. In AdaMatch, this threshold is
+relatively adjusted which is why it's called **relative confidence thresholding**. 
+
+For more details on these methods and to know how each of them contribute please refer to
+[the paper](https://arxiv.org/abs/2106.04732). 
+"""
+
+"""
+## Instantiate a Wide-ResNet-28-2
+
+The authors use a [WideResNet-28-2](https://arxiv.org/abs/1605.07146) for the dataset
+pairs we are using in this example. Note that the following model has a scaling layer
+inside it that scales the pixel values to [0, 1]. 
+"""
+
+"""shell
+wget -O wide_resnet.py -q https://git.io/Jnzzj
```

Comment:
We should definitely not be downloading arbitrary code to execute. Try to in-line the code in a function. If not possible, replace it with a generic Keras Application.

Reply:
Sure. I will inline the function. I was hesitant to do that to reduce the lines of code.